src/burnchains/burnchain.rs

```diff
@@ -305,322 +447,268 @@ impl Burnchain {
-                    info!("ACCEPT user burn support {}", &op.txid.to_hex());
-                    BurnDB::insert_user_burn(tx, op)
-                }
-            };
-
-        match_res
-            .map_err(burnchain_error::DBError)
+    fn check_transaction<'a>(tx: &mut BurnDBTx<'a>, burnchain: &Burnchain, block_header: &BurnchainBlockHeader, blockstack_op: &BlockstackOperationType) -> Result<(), burnchain_error> {
+        match blockstack_op {
+            BlockstackOperationType::LeaderKeyRegister(ref op) => {
+                op.check(burnchain, block_header, tx)
+                    .map_err(|e| {
+                          warn!("REJECTED({}) leader key register {} at {},{}: {:?}", op.block_height, &op.txid.to_hex(), op.block_height, op.vtxindex, &e);
+                          burnchain_error::OpError(e)
+                    })
+            },
+            BlockstackOperationType::LeaderBlockCommit(ref op) => {
+                op.check(burnchain, block_header, tx)
+                    .map_err(|e| {
+                          warn!("REJECTED({}) leader block commit {} at {},{}: {:?}", op.block_height, &op.txid.to_hex(), op.block_height, op.vtxindex, &e);
+                          burnchain_error::OpError(e)
+                    })
+            },
+            BlockstackOperationType::UserBurnSupport(ref op) => {
+                op.check(burnchain, block_header, tx)
+                    .map_err(|e| {
+                        warn!("REJECTED({}) user burn support {} at {},{}: {:?}", op.block_height, &op.txid.to_hex(), op.block_height, op.vtxindex, &e);
+                        burnchain_error::OpError(e)
+                    })
+            }
+        }
     }
 
-    /// Generate the list of blockstack operations that will be snapshotted.
-    /// Return the list of parsed blockstack operations whose check() method has returned true.
-    fn check_block<'a, A, K>(tx: &mut Transaction<'a>, burnchain: &Burnchain, block: &BurnchainBlock<A, K>) -> Result<Vec<BlockstackOperationType<A, K>>, burnchain_error>
-    where
-        A: Address,
-        K: PublicKey
-    {
-
-        debug!("Check block {} {}", block.block_height, &block.block_hash.to_hex());
-        let mut ret : Vec<BlockstackOperationType<A, K>> = vec![];
-
-        // classify and check each transaction
-        for i in 0..block.txs.len() {
-            match Burnchain::classify_transaction(block.block_height, &block.block_hash, &block.txs[i]) {
+    /// Filter out the burnchain block's transactions that could be blockstack transactions.
+    /// Return the ordered list of blockstack operations by vtxindex
+    fn get_blockstack_transactions(burnchain: &Burnchain, block: &BurnchainBlock, block_header: &BurnchainBlockHeader) -> Vec<BlockstackOperationType> {
+        debug!("Extract Blockstack transactions from block {} {}", block.block_height(), &block.block_hash().to_hex());
+        let mut ret = vec![];
```

Comment:
Something like this can be more simply and idiomatically implemented with a [filter_map](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter_map) and collect() -- you also shouldn't need to clone the blockstack_op

Reply:
Addressed in baa87c05